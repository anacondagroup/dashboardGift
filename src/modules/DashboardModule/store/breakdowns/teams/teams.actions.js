import {
  TEAMS_BREAKDOWN_LOAD_REQUEST,
  TEAMS_BREAKDOWN_LOAD_SUCCESS,
  TEAMS_BREAKDOWN_LOAD_FAILS,
  TEAMS_DOWNLOAD_REPORT_REQUEST,
  TEAMS_DOWNLOAD_REPORT_SUCCESS,
  TEAMS_DOWNLOAD_REPORT_FAIL,
} from './teams.types';

export const teamsLoadRequest = () => ({
  type: TEAMS_BREAKDOWN_LOAD_REQUEST,
});

export const teamsLoadSuccess = breakdown => ({
  type: TEAMS_BREAKDOWN_LOAD_SUCCESS,
  payload: breakdown,
});

export const teamsLoadFail = error => ({
  type: TEAMS_BREAKDOWN_LOAD_FAILS,
  payload: error,
});

export const teamsDownloadReportRequest = () => ({
  type: TEAMS_DOWNLOAD_REPORT_REQUEST,
});

export const teamsDownloadReportSuccess = () => ({
  type: TEAMS_DOWNLOAD_REPORT_SUCCESS,
});

export const teamsDownloadReportFail = () => ({
  type: TEAMS_DOWNLOAD_REPORT_FAIL,
});
