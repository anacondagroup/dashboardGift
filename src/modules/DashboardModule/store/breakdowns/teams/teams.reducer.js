import * as R from 'ramda';
import { renameKeys } from '@alycecom/utils';

import { initialBreakdownState } from '../../../helpers/breakdownSelectors.helpers';
import { setLoading, setLoaded } from '../../../../../helpers/lens.helpers';

import {
  TEAMS_BREAKDOWN_LOAD_REQUEST,
  TEAMS_BREAKDOWN_LOAD_SUCCESS,
  TEAMS_BREAKDOWN_LOAD_FAILS,
  TEAMS_DOWNLOAD_REPORT_REQUEST,
  TEAMS_DOWNLOAD_REPORT_SUCCESS,
  TEAMS_DOWNLOAD_REPORT_FAIL,
} from './teams.types';

export const initialState = {
  ...initialBreakdownState,
};

export const renameBreakdownKeys = renameKeys({
  gifts_sent: 'giftsSent',
  gifts_viewed: 'giftsViewed',
  gifts_accepted: 'giftsAccepted',
  meetings_booked: 'meetingsBooked',
});

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case TEAMS_BREAKDOWN_LOAD_REQUEST:
      return {
        ...state,
        ...setLoading(true, state),
      };
    case TEAMS_BREAKDOWN_LOAD_SUCCESS:
      return {
        ...state,
        ...R.compose(setLoading(false), setLoaded(true))(state),
        breakdown: action.payload.map(team => renameBreakdownKeys(team)),
      };
    case TEAMS_BREAKDOWN_LOAD_FAILS:
      return {
        ...state,
        ...setLoading(false, state),
        error: action.payload,
      };
    case TEAMS_DOWNLOAD_REPORT_REQUEST:
      return {
        ...state,
        isReportLoading: true,
      };
    case TEAMS_DOWNLOAD_REPORT_SUCCESS:
    case TEAMS_DOWNLOAD_REPORT_FAIL:
      return {
        ...state,
        isReportLoading: false,
      };
    default:
      return state;
  }
};
