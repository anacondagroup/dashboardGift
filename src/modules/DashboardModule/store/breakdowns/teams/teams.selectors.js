import * as R from 'ramda';
import { getUrlParams, updateSearch } from '@alycecom/modules';
import { renameKeys } from '@alycecom/utils';

import { getBreakdownsState } from '../breakdowns.selectors';
import { createSelectors } from '../../../helpers/breakdownSelectors.helpers';

const teamsApiParams = R.pick([
  'campaignId',
  'dateRangeFrom',
  'dateRangeTo',
  'teamsSearch',
  'teamsSort',
  'teamsDirection',
  'includeArchived',
]);

export const [
  getTeamsState,
  getTeamsBreakdown,
  getTeamsIsLoading,
  getTeamsReportIsLoading,
  getTeamsIsLoaded,
] = createSelectors(getBreakdownsState)('teams');

export const getApiCallQuery = R.compose(
  R.curryN(2, updateSearch)(''),
  renameKeys({
    campaignId: 'campaign_id',
    teamsSort: 'sort_column',
    teamsDirection: 'sort_direction',
    dateRangeFrom: 'date_range_from',
    dateRangeTo: 'date_range_to',
    teamsSearch: 'search',
    includeArchived: 'include_archived',
  }),
  teamsApiParams,
  getUrlParams,
);

export const getReportFileName = state => {
  let date = '';

  const urlParams = R.compose(teamsApiParams, getUrlParams)(state.router.location.search);

  if (urlParams.dateRangeFrom) {
    date = `-${urlParams.dateRangeFrom}`;
  }
  if (urlParams.dateRangeTo) {
    date += `-${urlParams.dateRangeTo}`;
  }

  return R.replace(/\./g, '_', `Teams-breakdown${date || ''}`).concat('.xlsx');
};
