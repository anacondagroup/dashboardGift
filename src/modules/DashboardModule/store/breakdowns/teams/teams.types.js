export const TEAMS_BREAKDOWN_LOAD_REQUEST = 'TEAMS_BREAKDOWN_LOAD_REQUEST';
export const TEAMS_BREAKDOWN_LOAD_SUCCESS = 'TEAMS_BREAKDOWN_LOAD_SUCCESS';
export const TEAMS_BREAKDOWN_LOAD_FAILS = 'TEAMS_BREAKDOWN_LOAD_FAILS';

export const TEAMS_DOWNLOAD_REPORT_REQUEST = 'TEAMS_DOWNLOAD_REPORT_REQUEST';
export const TEAMS_DOWNLOAD_REPORT_SUCCESS = 'TEAMS_DOWNLOAD_REPORT_SUCCESS';
export const TEAMS_DOWNLOAD_REPORT_FAIL = 'TEAMS_DOWNLOAD_REPORT_FAIL';
