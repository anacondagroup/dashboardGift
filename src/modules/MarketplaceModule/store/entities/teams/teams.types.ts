import { EntityId } from '@alycecom/utils';

export type TTeam = {
  id: EntityId;
  name: string;
};
