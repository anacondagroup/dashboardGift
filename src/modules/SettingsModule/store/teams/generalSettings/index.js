import * as selectors from './generalSettings.selectors';
import * as actions from './generalSettings.actions';
import * as schemas from './generalSettings.schemas';

export { selectors, actions, schemas };
