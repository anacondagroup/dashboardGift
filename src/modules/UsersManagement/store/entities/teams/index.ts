export { initialState, teams } from './teams.reducer';
export type { TTeamsState } from './teams.reducer';
export { teamsEpics } from './teams.epics';
export { getIsTeamsLoaded, getIsTeamsLoading, getTeamIds, getTeamsMap, getTeams } from './teams.selectors';
