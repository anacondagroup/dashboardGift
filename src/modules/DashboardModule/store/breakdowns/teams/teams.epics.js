import { ofType } from 'redux-observable';
import { catchError, map, mergeMap, withLatestFrom, tap, debounce, takeUntil } from 'rxjs/operators/index';
import { of, timer } from 'rxjs';

import { TEAMS_BREAKDOWN_LOAD_REQUEST, TEAMS_DOWNLOAD_REPORT_REQUEST } from './teams.types';
import { teamsLoadSuccess, teamsLoadFail, teamsDownloadReportSuccess, teamsDownloadReportFail } from './teams.actions';
import { getApiCallQuery, getReportFileName } from './teams.selectors';

export const loadTeamsBreakdownEpic = (action$, state$, { apiService }) =>
  action$.pipe(
    ofType(TEAMS_BREAKDOWN_LOAD_REQUEST),
    debounce(() => timer(300)),
    withLatestFrom(state$.pipe(map(state => getApiCallQuery(state.router.location.search)))),
    map(([, search]) => `/enterprise/dashboard/breakdown/teams?${search}`),
    mergeMap(url =>
      apiService.get(url).pipe(
        map(response => teamsLoadSuccess(response.teams)),
        catchError(response => of(teamsLoadFail(response))),
        takeUntil(action$.ofType(TEAMS_BREAKDOWN_LOAD_REQUEST)),
      ),
    ),
  );

export const downloadReportEpic = (action$, state$, { apiService, downloadFile }) =>
  action$.pipe(
    ofType(TEAMS_DOWNLOAD_REPORT_REQUEST),
    withLatestFrom(state$.pipe(map(state => getApiCallQuery(state.router.location.search)))),
    map(([, search]) => `/enterprise/report/team_breakdown?${search}`),
    mergeMap(url =>
      apiService.getFile(url).pipe(
        withLatestFrom(state$.pipe(map(getReportFileName))),
        tap(([blob, filename]) => {
          downloadFile(blob, filename);
        }),
        map(() => teamsDownloadReportSuccess()),
        catchError(() => teamsDownloadReportFail()),
      ),
    ),
  );

export const teamsEpic = [loadTeamsBreakdownEpic, downloadReportEpic];
