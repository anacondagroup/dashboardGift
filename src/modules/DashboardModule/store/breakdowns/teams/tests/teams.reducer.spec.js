import { reducer, initialState } from '../teams.reducer';
import { teamsLoadSuccess } from '../teams.actions';

describe('teams reducer', () => {
  describe('teamsLoadSuccess', () => {
    it('Should transform breakdown data to app format', () => {
      const actual = reducer(
        initialState,
        teamsLoadSuccess([
          {
            id: 1,
            name: 'WEX',
            gifts_sent: 0,
            gifts_viewed: 0,
            gifts_accepted: 0,
            meetings_booked: 0,
          },
        ]),
      );

      expect(actual).toMatchSnapshot();
    });
  });
});
