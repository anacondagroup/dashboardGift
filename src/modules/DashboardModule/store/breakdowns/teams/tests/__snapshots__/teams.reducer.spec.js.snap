// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`teams reducer teamsLoadSuccess Should transform breakdown data to app format 1`] = `
Object {
  "breakdown": Array [
    Object {
      "giftsAccepted": 0,
      "giftsSent": 0,
      "giftsViewed": 0,
      "id": 1,
      "meetingsBooked": 0,
      "name": "WEX",
    },
  ],
  "error": null,
  "isLoaded": true,
  "isLoading": false,
  "isReportLoading": false,
  "pagination": Object {
    "current_page": 1,
  },
}
`;
